app/tests/integration/conftest.py

```diff
@@ -0,0 +1,294 @@
+    ssh_key_path.touch(exist_ok=True)
+    ssh_key_path.write_text(keypair.private_key, encoding="utf-8")
+
+    yield types.SSHKey(keypair=keypair, private_key=ssh_key_path)
+
+    openstack_connection.delete_keypair(name=keypair.name)
+
+
+@pytest.fixture(scope="module", name="openstack_metadata")
+def openstack_metadata_fixture(
+    openstack_connection: Connection,
+    ssh_key: types.SSHKey,
+    network_name: str,
+    flavor_name: str,
+    cloud_name: str,
+) -> types.OpenstackMeta:
+    """A wrapper around openstack related info."""
+    return types.OpenstackMeta(
+        connection=openstack_connection,
+        cloud_name=cloud_name,
+        ssh_key=ssh_key,
+        network=network_name,
+        flavor=flavor_name,
+    )
+
+
+@pytest.fixture(scope="module", name="openstack_security_group")
+def openstack_security_group_fixture(openstack_connection: Connection, test_id: str):
+    """An ssh-connectable security group."""
+    security_group_name = f"github-runner-image-builder-test-security-group-{test_id}"
+    security_group: SecurityGroup | None = openstack_connection.get_security_group(
+        name_or_id=security_group_name
+    )
+    if not security_group:
+        security_group = openstack_connection.create_security_group(
+            name=security_group_name,
+            description="For servers managed by the github-runner-image-builder app.",
+        )
+        # For ping
+        openstack_connection.create_security_group_rule(
+            secgroup_name_or_id=security_group_name,
+            protocol="icmp",
+            direction="ingress",
+            ethertype="IPv4",
+        )
+        # For SSH
+        openstack_connection.create_security_group_rule(
+            # The code is not duplicated, this code is strictly for integration test which should
+            # not be imported from module or referenced from module.
+            # pylint: disable=R0801
+            secgroup_name_or_id=security_group_name,
+            port_range_min="22",
+            port_range_max="22",
+            protocol="tcp",
+            direction="ingress",
+            ethertype="IPv4",
+            # pylint: enable=R0801
+        )
+        # For tmate
+        openstack_connection.create_security_group_rule(
+            secgroup_name_or_id=security_group_name,
+            port_range_min="10022",
+            port_range_max="10022",
+            protocol="tcp",
+            direction="egress",
+            ethertype="IPv4",
+        )
+
+    yield security_group
+
+    openstack_connection.delete_security_group(security_group_name)
+
+
+@pytest.fixture(scope="module", name="proxy")
+def proxy_fixture(pytestconfig: pytest.Config) -> types.ProxyConfig:
+    """The environment proxy to pass on to the charm/testing model."""
+    # proxy has to be without http:// or https://  prefix
+    if pytestconfig.getoption("--proxy"):
+        proxy = pytestconfig.getoption("--proxy").removeprefix("http://").removeprefix("https://")
```

Comment:
added removeprefix because we are reusing the same INTEGRATION_TEST_ARGS as for the operator, and the operator expects the full proxy URL